api/src/main/java/org/infinispan/api/reactive/KeyValueStore.java

```diff
@@ -0,0 +1,50 @@
+package org.infinispan.api.reactive;
+
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * A Reactive Key Value Store provides a highly concurrent and distributed data structure, non blocking and using
+ * reactive streams.
+ * <p>
+ * TODO: Add more javadoc and examples
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @see <a href="http://www.infinispan.org">Infinispan documentation</a>
+ * @since 10.0
+ */
+public interface KeyValueStore<K, V> {
+
+   CompletionStage<V> get(K key);
+
+   CompletionStage<Void> insert(K key, V value);
+
+   CompletionStage<Void> save(K id, V value);
+
+   CompletionStage<V> getAndSave(K key, V value);
+
+   CompletionStage<Void> delete(K key);
+
+   CompletionStage<V> getAndDelete(K key);
+
+   Publisher<K> keys();
+
+   Publisher<Map.Entry<K, V>> entries();
```

Comment:
Doubly embedded type parameters can sometimes become an issue with subclasses. I don't know the rules exactly but @wburns had to fight a few of those with the stream stuff. Having said that, this can be quickly detected with appropriate testing, e.g, if Dog extends Animal, can I store my dog inside a KeyValueStore that takes Animal as value?

Reply:
Can we leave it this way from now and rework ?

Reply:
It should probably be `Publisher<? extends Map.Entry<K, V>>` to avoid unchecked casts in the implementation

Reply:
ok @danberindei